mirrord/sip/src/lib.rs

```diff
@@ -287,15 +304,55 @@ mod main {
         }
     }
 
+    /// Add the given `path` to excluded paths.
+    fn add_to_file_exclude(path: &str) {
+        env::set_var(
+            EXCLUDE_ENV_VAR_NAME,
+            env::var(EXCLUDE_ENV_VAR_NAME)
+                .map(|paths| paths + ";")
+                .unwrap_or(String::new())
+                + path,
+        )
+    }
+
+    /// Get a path to a temp dir that is excluded from FS hooks and is also saved in an env var so
+    /// that child processes also use the same one and exclude it from FS hooks.
+    fn get_tmp_dir() -> Result<PathBuf> {
+        env::var(TMP_DIR_ENV_VAR_NAME)
+            .map(PathBuf::from)
+            .or_else(|_err| {
+                let tmp_dir = temp_dir().join("mirrord-bin");
+                let tmp_dir_string = tmp_dir
+                    .to_str()
+                    .ok_or(UnlikelyError(
+                        "Failed to convert path to string".to_string(),
+                    ))?
+                    .to_string();
+                env::set_var(TMP_DIR_ENV_VAR_NAME, &tmp_dir_string);
+                add_to_file_exclude(&(tmp_dir_string + "/.*$"));
+                Ok(tmp_dir)
+            })
+    }
+
     /// Check if the file that the user wants to execute is a SIP protected binary (or a script
-    /// starting with a shebang that leads to a SIP protected binary). If it is, create a
-    /// non-protected version of the file and return the path to it. If it is not, the original
-    /// path is copied and returned.
-    pub fn sip_patch(binary_path: &str) -> Result<String> {
+    /// starting with a shebang that leads to a SIP protected binary).
+    /// If it is, create a non-protected version of the file and return `Ok(Some(patched_path)`.
+    /// If it is not, `Ok(None)`.
+    /// Propagate errors.
+    pub fn sip_patch(binary_path: &str) -> Result<Option<String>> {
+        // Important: do also if the binary has no SIP, so that a temp directory is set, added to
+        // ignored files, and saved in an environment variable, so that if the binary execs SIP
+        // binaries the patched versions are created in a directory that is already added to the
+        // excluded files. (It's currently not possible to add excluded paths when the layer is
+        // already running, and the first call to this function is from the cli.)
+        let tmp_dir = get_tmp_dir()?;
+
         if let SipStatus::SomeSIP(path, shebang_target) = get_sip_status(&binary_path)? {
-            return patch_some_sip(&path, shebang_target);
+            debug!("Using temp dir: {:?} for sip patches", &tmp_dir);
```

Comment:
```suggestion
            trace!("Using temp dir: {:?} for sip patches", &tmp_dir);
```

Let's try leaving everything useful only to us as `trace` when it comes time to merge.

Reply:
Done. What's `debug` for then?

Reply:
Hmm I would leave `debug` for when we're in the middle of developing something. I'm trying to use it like that, then putting everything back to `trace` once I'm done.

Just an idea though, we have not really discussed this type of stuff (only that we should avoid leaking higher tracing levels to the user with information they don't need).